src/box/txn.c

```diff
@@ -1201,15 +1195,17 @@ txn_commit_impl(struct txn *txn, enum txn_commit_wait_mode wait_mode)
 	 * txn not waiting for anything.
 	 */
 	if (txn_has_flag(txn, TXN_WAIT_SYNC) &&
-	    txn_add_limbo_entry(txn, req) != 0) {
+	    txn_add_limbo_entry(txn, req, wait_mode) != 0) {
 		goto rollback_abort;
 	}
 	fiber_set_txn(fiber(), NULL);
 	if (journal_write_submit(req) != 0) {
 		fiber_set_txn(fiber(), txn);
 		goto rollback_io;
 	}
-	if (wait_mode != TXN_COMMIT_WAIT_MODE_COMPLETE) {
+	if ((!txn_has_flag(txn, TXN_WAIT_SYNC) ||
+	     txn_limbo.owner_id != instance_id) &&
+	    wait_mode != TXN_COMMIT_WAIT_MODE_COMPLETE) {
 		if (txn_has_flag(txn, TXN_IS_DONE))
 			goto finish_done;
 		txn->fiber = NULL;
```

Comment:
Two lines below there's an explicit yield in the `none` commit mode, if tx is synchronous. 

```C
	while (!req->is_complete)
		fiber_yield();
```

Why don't we explicitly forbid using sync tx with `none` mode?

Reply:
Fixed: after #10628 we can immediately return here.